### PR TITLE
Allow passing in a sinon stub or jest mock as the replacement function

### PR DESCRIPTION
**Pull request by thomaux**
This PR allows you to pass in a sinon stub or jest mock function as the replacement function:

```javascript
const jestMock = jest.fn().mockReturnValue('message');
awsMock.mock('DynamoDB', 'getItem', jestMock);
const db = new AWS.DynamoDB();
db.getItem({}, function(err, data){
      expect(data).toEqual('message')
      expect(jestMock).toHaveBeenCalledTimes(1);
});
```

Although the docs stated you could pass in a sinon spy, the spy was not functioning as an actual replacement function. I added a wrapper function that detects if the replacement parameter is a function and whether it is a sinon stub or a jest mock. If so, the stub/mock is wrapped as a fully functional replacement function that honours user callbacks (as shown in the example above).

This is the relevant code change:

```javascript
function wrapTestStubReplaceFn(replace) {
  if (typeof replace !== 'function' || !(replace._isMockFunction || replace.isSinonProxy)) {
    return replace;
  }

  return (params, cb) => {
    let result;
    try {
      result = replace(params);
      if (typeof result.then === 'function') {
        result.then(val => cb(undefined, val), err => cb(err));
      } else {
        cb(undefined, result);
      }
    } catch (err) {
      cb(err);
    }
  };
}
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -113,6 +113,29 @@ function mockService(service) {
...
+ */
+function wrapTestStubReplaceFn(replace) {
+  if (typeof replace !== 'function' || !(replace._isMockFunction || replace.isSinonProxy)) {
+    return replace;
+  }
+
+  return (params, cb) => {
+    let result;
+    try {
+      result = replace(params);
```
Comment (abetomo) [inline index.js:127]: It is better to use 'const' instead of 'let'.

```js
const result = replace(params);
```

File: test/index.test.js [hunk 2]
```diff
@@ -583,6 +585,76 @@ test('AWS.mock function should mock AWS service and method on the service', func
...
+    db.getItem({}, function(err, data){
+      st.equal(data, 'message');
+      st.equal(jestMock.mock.calls.length, 1);
+      st.end();
+    });
+  });
+
+  t.test('mock function replaces method with a jest mock and fails successfully', function(st) {
+    const jestMock = jest.fn().mockImplementation(() => {
+        throw new Error('something went wrong')
```
Comment (abetomo) [inline test/index.test.js:636]: There is a lot of indentation.

**Review by abetomo — COMMENTED** [review]
Great work!

I left a few suggestions, please take a look?

File: index.js (hunk 1 above)
Comment (nelsonic) [reply index.js]: Updated on branch. Thanks @abetomo

**Review by nelsonic — DISMISSED** [review]
@thomaux this is a great addition for the people using `jest`. Thank you! 👍 
I just wish there weren't so many dependencies ... 🙄 
Hopefully there won't be too much noise and @snyk will help with keeping them up-to-date. 🤞 
Thanks again. 🙌

**Review by abetomo — APPROVED** [review]
👍

**Review by nelsonic — APPROVED** [review]
Apologies for dismissing the review/approval. Unintentional. 🙄 
Just wanted to bump the package version in the PR so it's self-contained. 📦

**Comment by nelsonic** [comment]
@thomaux FYI: `aws-sdk-mock@5.5.0` published to NPM contains this update. 📦 
Thanks again. ☀️

**Comment by thomaux** [comment]
@nelsonic and @abetomo great, thanks for the review and feedback!

**Comment by pascaljung** [comment]
Unfortunately, this change breaks our unit tests since we quite often use the following pattern:

```js
let origMock = jest.fn((p, cb) => { cb(null, {}); });
AWSMock.mock('S3', methodName, origMock);
```

Can we maybe change the implementation to check if the passed mock function has a callback set as second parameter and use that to resolve the promise? @nelsonic what do you think?

**Pull request merged**